src/module.c

```diff
@@ -3655,6 +3674,130 @@ void moduleReleaseGIL(void) {
+ *  - REDISMODULE_NOTIFY_EVICTED: Eviction events
+ *  - REDISMODULE_NOTIFY_ALL: All events
+ * 
+ * We do not distinguish between key events and keyspace events, and it is up
+ * to the module to filter the actions taken based on the key. 
+ * 
+ * The subscriber signature is:
+ * 
+ *   int (*RedisModuleNotificationFunc) (RedisModuleCtx *ctx, int type, 
+ *                                       const char *event, 
+ *                                       RedisModuleString *key);
+ * 
+ * `type` is the event type bit, that must match the mask given at registration
+ * time. The event string is the actual command being executed, and key is the
+ * relevant Redis key. 
+ * 
+ * Notification callback gets executed with a redis context that can not be
+ * used to send anything to the client, and has the db number where the event
+ * occured as its selected db number.
+ * 
+ * Notice that it is not necessary to enable norifications in redis.conf for 
+ * module notifications to work. 
+ * 
+ * Warning: the notification callbacks are performed in a synchronous manner,
+ * so notification callbacks must to be fast, or they would slow Redis down. 
+ * If you need to take long actions, use threads to offload them.
+ * 
+ * See https://redis.io/topics/notifications for more information. 
+ */
+int RM_SubscribeToKeyspaceEvents(RedisModuleCtx *ctx, int types, RedisModuleNotificationFunc callback) {
+    RedisModuleKeyspaceSubscriber *sub = zmalloc(sizeof(*sub));
+    sub->module = ctx->module;
+    sub->event_mask = types;
+    sub->notify_callback = callback;
+    sub->active = 0;
+
+    /* Let the notification system know that modules are interested in notifications */
+    server.notify_keyspace_events |= NOTIFY_MODULE;
+    listAddNodeTail(moduleKeyspaceSubscribers, sub);
+    return REDISMODULE_OK;
+
+}
+
+/* Dispatcher for keyspace notifications to module subscriber functions. 
+ * This gets called  only if at least one module requested to be notified on
+ * keyspace notifications */
+void moduleNotifyKeyspaceEvent(int type, const char *event, robj *key, int dbid) {
+    listIter li;
+    listNode *ln;
+
+    listRewind(moduleKeyspaceSubscribers,&li);
+
+    /* Remove irrelevant flags from the type mask */
+    type &= ~(NOTIFY_KEYEVENT | NOTIFY_KEYSPACE);
+
+    /* Setup a fake client, so we can have proper db selection when performing
+     * actions. We use one client for all handlers, writing to it will crash */
+    client *c = createClient(-1);
+    c->flags |= CLIENT_MODULE;
+    
+    while((ln = listNext(&li))) {
+        RedisModuleKeyspaceSubscriber *sub = ln->value;
+        /* Only notify subscribers on events matching they registration,
+         * and avoid subscribers triggering themselves */
+        if ((sub->event_mask & type) && sub->active == 0) {
+            RedisModuleCtx ctx = REDISMODULE_CTX_INIT;
+            ctx.module = sub->module;
+            selectDb(c, dbid);
+            ctx.client = c;
```

Comment:
note that I'm using the same client for all contexts. I didn't want to use the same context, even though it will save a shitload of CPU time in these tight loops, because of auto memory constraints, etc. But if you feel that it's safe perhaps we can just recycle the module context completely.